crates/core/c8y_api/src/smartrest/smartrest_serializer.rs

```diff
@@ -96,19 +96,67 @@ fn succeed_static_operation(
 /// rules will be applied. For example, `a,field "with" quotes` will be converted to
 /// `a,"field ""with"" quotes"` before being appended to the output of this function.
 ///
+/// If the message size turns out to be bigger than maximum cumulocity message size, it will be truncated to fit in the
+/// limit.
+///
 /// # Errors
 /// This will return an error if the payload is a CSV with multiple records, or an empty CSV.
 pub fn succeed_operation(
     template_id: &str,
     operation: &str,
     reason: impl Into<TextOrCsv>,
 ) -> Result<String, SmartRestSerializerError> {
-    let mut wtr = csv::Writer::from_writer(vec![]);
-    // Serialization can fail for CSV, but not for text
-    wtr.serialize((template_id, operation, reason.into()))?;
-    let mut output = wtr.into_inner().unwrap();
-    output.pop();
-    Ok(String::from_utf8(output)?)
+    let reason: TextOrCsv = reason.into();
+
+    let result = {
+        let mut wtr = csv::Writer::from_writer(vec![]);
+        wtr.serialize((template_id, operation, &reason))?;
+        let mut vec = wtr.into_inner().unwrap();
+        // remove newline character
+        vec.pop();
+        String::from_utf8(vec)?
+    };
+
+    if result.len() <= super::message::MAX_PAYLOAD_LIMIT_IN_BYTES {
+        return Ok(result);
+    }
+
+    // payload too big, need to trim
+    let prefix = super::csv::fields_to_csv_string(&[template_id, operation]);
+
+    // 3 extra characters: 1 for comma, 2 for surrounding quotes
+    let mut max_result_limit = super::message::MAX_PAYLOAD_LIMIT_IN_BYTES - prefix.len() - 3;
+
+    // escaping can add additional characters so that the field will be too large to fit in the
+    // message; if so we need to trim it after escaping, making sure we don't screw up bespoke
+    // smartrest escape sequences
+    let result = {
+        let mut wtr = csv::WriterBuilder::new()
+            .quote_style(csv::QuoteStyle::Always)
+            .from_writer(vec![]);
+        wtr.serialize(reason)?;
+        let mut vec = wtr.into_inner().unwrap();
+
+        // remove newline character
+        vec.pop();
+
+        // remove outer quotes, added back after trimming
+        let reason = std::str::from_utf8(&vec).unwrap();
+        let reason = reason.strip_prefix('"').unwrap_or(reason);
+        let reason = reason.strip_suffix('"').unwrap_or(reason);
+
+        // if we'd cut across an escaped " character, move trim point 1 char back to omit it
+        if &reason[max_result_limit - 1..=max_result_limit] == r#""""# {
+            max_result_limit -= 1;
+        }
+        let trimmed_reason = &reason[..max_result_limit];
+
+        format!("{prefix},\"{trimmed_reason}\"")
```

Comment:
It would be good to add a hint that the reason has been truncated as suggested by @reubenmiller: https://github.com/thin-edge/thin-edge.io/pull/3179#issuecomment-2413316148

```suggestion
        format!("{prefix},\"{trimmed_reason} ...<truncated>\"")
```